csi_camera.cpp

```diff
@@ -0,0 +1,451 @@
+            }
+
+            // Check bus for messages
+            msg = gst_bus_pop(bus);
+            if (msg != nullptr) {
+                catch_pipeline();
+                gst_message_unref(msg);
+                msg = nullptr;
+            }
+
+            return vec;
+        }
+
+        // UTILS 
+
+        std::string create_pipeline() {
+            std::ostringstream oss;
+
+            oss << input_source << " sensor_id=" << video_path
+                << " ! " << DEFAULT_INPUT_FORMAT
+                << ",width=" << std::to_string(width_px)
+                << ",height=" << std::to_string(height_px)
+                << ",framerate=" << std::to_string(frame_rate)
+                << "/1 ! nvvidconv flip-method=" << DEFAULT_INPUT_FLIP_METHOD
+                << " ! " << DEFAULT_OUTPUT_FORMAT
+                << ",width=" << std::to_string(DEFAULT_OUTPUT_WIDTH)
+                << ",height=" << std::to_string(DEFAULT_OUTPUT_HEIGHT)
+                << " ! " << output_encoder
+                << " ! " << DEFAULT_OUTPUT_MIMETYPE
+                << " ! appsink name=appsink0 max-buffers=1";
+
+            return oss.str();
+        }
+
+        std::vector<unsigned char> buff_to_vec(GstBuffer *buff) {
+            // Get the size of the buffer
+            size_t bufferSize = gst_buffer_get_size(buffer);
+
+            // Create a vector with the same size as the buffer
+            std::vector<unsigned char> vec(bufferSize);
+
+            // Copy the buffer data to the vector
+            GstMapInfo map;
+            gst_buffer_map(buffer, &map, GST_MAP_READ);
+            memcpy(vec.data(), map.data, bufferSize);
+            gst_buffer_unmap(buffer, &map);
+
+            return vec;
+        }
+        
+        std::vector<unsigned char> get_test_image() {
+            std::string test_image_path = "./etc/viam-logo.jpeg";
+            
+            // Create filestream
+            std::ifstream file(test_image_path, std::ios::binary);
+            if (!file) {
+                std::cout << "ERROR: could not open file" << std::endl;
+                return {};
+            }
+
+            // Read the file contents into a vector
+            std::vector<unsigned char> bytes(std::istreambuf_iterator<char>(file), {});
+
+            return bytes;
+        }
+
+        // GETTERS
+
+        std::string get_video_path() const {
+            return video_path;
+        }
+
+        int get_width_px() const {
+            return width_px;
+        }
+
+        int get_height_px() const {
+            return height_px;
+        }
+
+        int get_frame_rate() const {
+            return frame_rate;
+        }
+
+        bool is_debug() const {
+            return debug;
+        }
+
+        GstElement* get_appsink() const {
+            return appsink;
+        }
+        
+        GstElement* get_pipeline() const {
+            return pipeline;
+        }
+
+        GstBus* get_bus() const {
+            return bus;
+        }
+};
```

Comment:
Does a modular resource not have to implement close()?

Reply:
There is no explicit "close" api AFAIK.

From docs:

- During shutdown, a RemoveResource() is sent for every resource handled by a module, instructing the module to close and remove it.
- Lastly, the module process itself is stopped by the parent, before the parent itself terminates.

Added deconstructor to make sure GST pipeline closes down when the resource is removed.

Reply:
Ok cool, if resources support close by implementing a destructor, then that sounds fine